### ISSUE TITLE
Using %%time gives confusing traceback at error

### ISSUE THREAD

**Issue by joelmo** [ipython/ipython#9266]
Without magic:

```
foo
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-26-d3b07384d113> in <module>()
----> 1 foo

NameError: name 'foo' is not defined
```

With magic:

```
%%time
foo
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-27-81319e04df6a> in <module>()
----> 1 get_ipython().run_cell_magic(u'time', u'', u'foo')

/nix/store/qlvl1wqwcc2gx8mjqpb8fxbzgv0lyayc-python2.7-ipython-4.0.0/lib/python2.7/site-packages/IPython/core/interactiveshell.pyc in run_cell_magic(self, magic_name, line, cell)
   2291             magic_arg_s = self.var_expand(line, stack_depth)
   2292             with self.builtin_trap:
-> 2293                 result = fn(magic_arg_s, cell)
   2294             return result
   2295 

<decorator-gen-60> in time(self, line, cell, local_ns)

/nix/store/qlvl1wqwcc2gx8mjqpb8fxbzgv0lyayc-python2.7-ipython-4.0.0/lib/python2.7/site-packages/IPython/core/magic.pyc in <lambda>(f, *a, **k)
    191     # but it's overkill for just that one bit of state.
    192     def magic_deco(arg):
--> 193         call = lambda f, *a, **k: f(*a, **k)
    194 
    195         if callable(arg):

/nix/store/qlvl1wqwcc2gx8mjqpb8fxbzgv0lyayc-python2.7-ipython-4.0.0/lib/python2.7/site-packages/IPython/core/magics/execution.pyc in time(self, line, cell, local_ns)
   1161         if mode=='eval':
   1162             st = clock2()
-> 1163             out = eval(code, glob, local_ns)
   1164             end = clock2()
   1165         else:

<timed eval> in <module>()

NameError: name 'foo' is not defined
```

Is it possible to have same output in these cases, and maybe optionally turn on extra traceback when using %%time?

```
{'commit_hash': u'f534027',
 'commit_source': 'installation',
 'default_encoding': 'UTF-8',
 'ipython_path': '/nix/store/qlvl1wqwcc2gx8mjqpb8fxbzgv0lyayc-python2.7-ipython-4.0.0/lib/python2.7/site-packages/IPython',
 'ipython_version': '4.0.0',
 'os_name': 'posix',
 'platform': 'Linux-4.1.15-x86_64-with-glibc2.2.5',
 'sys_executable': '/nix/store/d945ibfx9x185xf04b890y4f9g3cbb63-python-2.7.11/bin/python2.7',
 'sys_platform': 'linux2',
 'sys_version': '2.7.11 (default, Dec  5 2015, 19:47:20) \n[GCC 4.9.3]'}
```

**Comment by takluyver** [comment]
I guess that time and timeit could catch exceptions executing the code and call `self.shell.showtraceback()` themselves; that should probably work.

**Comment by srinivasreddy** [comment]
PR has been raised  https://github.com/ipython/ipython/pull/10108

**Comment by takluyver** [comment]
Fixed by #10108
